### PR TITLE
ENH: Update image utility output path behaviour

### PR DESCRIPTION
**Pull request by josephmje**

### CODE REVIEW

File: dmriprep/utils/images.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    mask_data = mask_img.get_fdata(dtype='float32')
+
+    median_signal = np.median(data[mask_data > 0, ...], axis=0)
+    rescaled_data = 1000 * data / median_signal
+    hdr = img.header.copy()
+    nb.Nifti1Image(rescaled_data, img.affine, hdr).to_filename(out_path)
+    return out_path
+
+
+def median(in_file, out_path=None):
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
def median(in_file, dtype=None, out_path=None):
```

File: dmriprep/utils/images.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            in_file, suffix='_b0ref', use_ext=True)
+
+    img = nb.load(in_file)
+    if img.dataobj.ndim == 3:
+        return in_file
+    if img.shape[-1] == 1:
+        nb.squeeze_image(img).to_filename(out_path)
+        return out_path
+
+    median_data = np.median(img.get_fdata(dtype='float32'),
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
    median_data = np.median(img.get_fdata(dtype=dtype),
```

File: dmriprep/utils/images.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                            axis=-1)
+
+    hdr = img.header.copy()
+    hdr.set_xyzt_units('mm')
+    hdr.set_data_dtype(np.float32)
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
    if dtype is not None:
        hdr.set_data_dtype(dtype)
``

**Comment by josephmje** [comment]
@oesteban should `dtype` be added as an argument to the other functions as well?

File: dmriprep/utils/images.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+
+def extract_b0(in_file, b0_ixs, out_path=None):
+    """Extract the *b0* volumes from a DWI dataset."""
+    if out_path is None:
+        out_path = fname_presuffix(
+            in_file, suffix='_b0', use_ext=True)
+
+    img = nb.load(in_file)
+    data = img.get_fdata(dtype='float32')
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
    data = img.get_fdata(dtype='float32')
```

File: dmriprep/utils/images.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+    b0 = data[..., b0_ixs]
+
+    hdr = img.header.copy()
+    hdr.set_data_shape(b0.shape)
+    hdr.set_xyzt_units('mm')
+    hdr.set_data_dtype(np.float32)
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
```

File: dmriprep/utils/images.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    nb.Nifti1Image(b0, img.affine, hdr).to_filename(out_path)
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
    nb.Nifti1Image(b0.astype(hdr.get_data_dtype()), img.affine, hdr).to_filename(out_path)
```

File: dmriprep/utils/images.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    """Rescale the input volumes using the median signal intensity."""
+    if out_path is None:
+        out_path = fname_presuffix(
+            in_file, suffix='_rescaled_b0', use_ext=True)
+
+    img = nb.load(in_file)
+    if img.dataobj.ndim == 3:
+        return in_file
+
+    data = img.get_fdata(dtype='float32')
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
    data = img.get_fdata()
```

File: dmriprep/utils/images.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+    mask_img = nb.load(mask_file)
+    mask_data = mask_img.get_fdata(dtype='float32')
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
    mask_data = mask_img.get_fdata()
```

File: dmriprep/utils/images.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+    median_signal = np.median(data[mask_data > 0, ...], axis=0)
+    rescaled_data = 1000 * data / median_signal
+    hdr = img.header.copy()
+    nb.Nifti1Image(rescaled_data, img.affine, hdr).to_filename(out_path)
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
    nb.Nifti1Image(rescaled_data.astype(hdr.get_data_dtype()), img.affine, hdr).to_filename(out_path)
```

File: dmriprep/utils/images.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+        return out_path
+
+    median_data = np.median(img.get_fdata(dtype='float32'),
+                            axis=-1)
+
+    hdr = img.header.copy()
+    hdr.set_xyzt_units('mm')
+    if dtype is not None:
+        hdr.set_data_dtype(dtype)
+    nb.Nifti1Image(median_data, img.affine, hdr).to_filename(out_path)
```
Comment (oesteban) [inline dmriprep/utils/images.py]:
```suggestion
    else:
        dtype = hdr.get_data_dtype()
    nb.Nifti1Image(median_data.astype(dtype), img.affine, hdr).to_filename(out_path)
```

**Comment by josephmje** [comment]
We might need to change the version of nipype that's pinned. The new version is getting `TraitType` from `traits.api` but the older one imported from `traits.trait_handlers`.

**Comment by oesteban** [comment]
Could you rebase?

File: dmriprep/interfaces/images.py [hunk 11]
```diff
@@ -91,55 +85,27 @@ class RescaleB0(SimpleInterface):
...
+            suffix='_ref',
+            use_ext=True,
+            newpath=str(Path(runtime.cwd).absolute())
+        )
+
         self._results['out_b0s'] = rescale_b0(
             self.inputs.in_file,
-            self.inputs.mask_file,
-            newpath=runtime.cwd
+            self.inputs.mask_file, out_b0s
```
Comment (oesteban) [inline dmriprep/interfaces/images.py:105]: I think this is what introduced #90 - the mask and the b0s are swapped.

Comment (josephmje) [reply dmriprep/interfaces/images.py:105]: it failed at the very first step where the b0s get extracted

Comment (oesteban) [reply dmriprep/interfaces/images.py:105]: yup I realized after commenting. I'm still going through this - I might have found a solution.

**Pull request merged**